### PR TITLE
[WIP]: Remove html-to-jsx

### PR DESCRIPTION
**Pull request by ChristopherBiscardi**
Long story short, we still need html-to-jsx applied to the result of the html producing remark plugins that gatsby uses. Our first problem is how to identify JSX, Fragments of JSX, and html from each other. Then we can attempt to apply a hast plugin to the resulting html AST to convert to JSX.

This is a WIP that includes some initial tests. Ideas welcome.

The html-to-jsx plugin is also preventing us from upgrading projects to the React alpha afaict.

### CODE REVIEW

**Comment by ChristopherBiscardi** [comment]
current output for style attribute is `style="{{\"display\":\"block\"}}"`
desired is `style={{"display":"block"}}`

Otherwise this approach of using a babel plugin seems to work the best. Need to clean up the PR, fix the final output and ship it.

**Comment by ChristopherBiscardi** [comment]
style output is fixed. Some `class` attributes make it through still as well as plugins like katex creating jsx-incompatible output like:

```html
<annotation encoding="application/x-tex">
  c = \pm\sqrt{a ^ (2 + b) ^ 2}
</annotation>
```

**Comment by ChristopherBiscardi** [comment]
maybe we should be escaping jsx-incompatible content in the children? idk. Perhaps we should add "patches" that look for plugins that we know will generate incompatible output like katex. (find the annotations with parent blocks that have a katex class or something)

**Comment by ChristopherBiscardi** [comment]
srcset needs to transform to `srcSet` too (with the class -> className, for -> htmlFor object)

**Comment by ChristopherBiscardi** [comment]
# Current MDX Release

image

# This PR

image

**Comment by ChristopherBiscardi** [comment]
notably the code sample is all on one line? I don't know why. Need to dig a bit.

**Comment by wooorm** [comment]
@ChristopherBiscardi Which plugins expose HTML instead of HAST (I think that’s what you’re talking about here, right?)? I’d like to fix that at the source as well.

**Comment by ChristopherBiscardi** [comment]
@wooorm All of the [gatsby-remark-*](https://www.gatsbyjs.org/plugins/?=gatsby-remark) plugins, such as [gatsby-remark-prism](https://github.com/gatsbyjs/gatsby/blob/768aaa03206ac0b8d50a9d816d7e92ea176d3abc/packages/gatsby-remark-prismjs/src/index.js#L64) which injects html into the value like:

```js
    node.value = ``
    + `<div class="gatsby-highlight" data-language="${languageName}">`
    +   `<pre${numLinesStyle} class="${className}${numLinesClass}">`
    +     `<code class="${className}">`
    +       `${highlightCode(language, node.value, highlightLines)}`
    +     `</code>`
    +     `${numLinesNumber}`
    +   `</pre>`
```

**Comment by ChristopherBiscardi** [comment]
code sample is all on one line because html vs jsx differences in behavior relating to newline :duh:. Sometimes I just need a nap lol

**Comment by wooorm** [comment]
Ahh OK! Hmm, I believe all normal remark-* packages integrate with HAST/rehype, so that would also be possible I guess, to make sure gatsby-remark-* packages do that as well, but maybe more work?

**Comment by ChristopherBiscardi** [comment]
@wooorm yeah, now that you've brought it up I'll probably go through the gatsby packages and submit PRs to convert the most popular ones to HAST output, but I still have to implement the fallbacks because there's a long tail of plugins that aren't appropriately written.

Do you know of a good way to "template" HAST from an html string? Right now I'm thinking this approach might be most maintainable for gatsby-remark-* plugins that output a lot of HTML.

```js
    node = remark.parse(`<div class="gatsby-highlight" data-language="${languageName}">
  <pre${numLinesStyle} class="${className}${numLinesClass}">`
    <code class="${className}">`
    ${highlightCode(language, node.value, highlightLines)}
    </code>
    ${numLinesNumber}
  </pre>
</div>`)
```

**Comment by ChristopherBiscardi** [comment]
Going to merge this more-or-less as-is. The biggest problem right now is that the gatsby-remark-* plugins insert raw HTML and not HAST (HAST would be transformed by @mdx-js/mdx in the core pipeline, whereas we are transforming *after* the full compilation using babel).

@wooorm yeah, iirc I tried rehype-raw on the whole tree and it choked on the mdx `import` nodes. I think a second attempt at this PR could be a custom rehype-raw-lax, where if hast-util-raw chokes when trying to parse, then it just backs out and passes the node through as is. The other problem is that I've seen rehype-parse, etc downcase JSX in my tests. So it'll just think everything is fine and we end up with malformatted JSX :/

File: packages/gatsby-mdx/utils/babel-plugin-html-attr-to-jsx-attr.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  }
+};
+
+module.exports = function attrs() {
+  return {
+    visitor: {
+      JSXElement: function(path) {
+        path.traverse(jsxAttributeFromHTMLAttributeVisitor);
+
+        if (
```
Comment (ChristopherBiscardi) [inline packages/gatsby-mdx/utils/babel-plugin-html-attr-to-jsx-attr.js]: This is an example patch for the gatsby-remark-katex plugin. It replaces symbols that represent JSXExpressions with their html attributes instead in <annotation>. We might have to do a few of these in the meantime while we update gatsby-remark-* plugins to insert HAST and not raw html

**Pull request merged**